File: memoized/kotlin/canSum.kt
fun canSum(targetSum: Int, numbers: Array<Int>, memo: HashMap<Int, Boolean> = HashMap()): Boolean {
    if (memo.containsKey(targetSum)) return memo[targetSum] ?: false
    if (targetSum < 0) return false;
    if (targetSum == 0) return true;

    numbers.forEach {
        memo[targetSum] = canSum(targetSum - it, numbers, memo)
        if (memo[targetSum] ?: false) return true
    }

    memo[targetSum] = false
    return false
}

fun main(vararg args: String): Unit {
    println(canSum(args[0].toInt(), arrayOf<Int>(14, 7)))
}

File: memoized/javascript/canSum.js
const canSum = (target, numbers, memo = {}) => {
    if (target in memo) return memo[target]
    if (target < 0) return false;
    if (target === 0) return true;

    for (let i of numbers) {
        memo[target] = canSum(target - i, numbers, memo)
        if (memo[target]) return true
    }

    return memo[target] = false
}

console.log(canSum(3553, [7, 14, 28]))